### PR TITLE
Parallelization refactor

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Summary

The goal of this PR is to:
- refactor the User Guide parallelization to differentiate between TBB and MPI and include OpenCL stuff
- add parallelization section to CmdStan manual (we had one but it was bad and confusing and not informative)

This is a WIP but I need feedback.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Rok Češnovar, Erik Štrumbelj, Univ. of Ljubljana

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/cmdstan-guide/parallelization.Rmd [hunk 1]
```diff
@@ -1,37 +1,127 @@
...
+
+Generally, this number should not exceed the number of available cores.
+If this variable isn't set, then the program will run with only one thread.
+
+Once `STAN_NUM_THREADS` is set, run a model as normal:
+```
+./model sample data file=data.csv ...
+```
+
+## Multi-processing with MPI
```
Comment (rok-cesnovar) [inline src/cmdstan-guide/parallelization.Rmd:54]: @yizhang-yiz mind skimming over this subsection? Thanks!

**Comment by rok-cesnovar** [comment]
@bbbales2 can you take a quick look and give me your thoughts on this?

I am trying to primarily add OpenCL stuff but also differentiate a bit more between multi-thread and processing as I feel this confuses users a lot.

File: src/cmdstan-guide/parallelization.Rmd [hunk 2]
```diff
@@ -1,37 +1,127 @@
...
+```
+The model is then compiled as normal:
+```
+make path/to/model
+```
+### Running
+
+The Stan model compiled with `STAN_MPI` is then run using an MPI launcher (typically `mpiexec` or `mpirun`).
+The launcher is supplied the path to the built executable and the number of processes to start (`-n`).
+
```
Comment (yizhang-yiz) [inline src/cmdstan-guide/parallelization.Rmd:91]: MPI Standard suggests `mpiexec`, which takes `-n`. `mpirun` is vendor wrapper, takes `-np` as nproc flag.

Comment (rok-cesnovar) [reply src/cmdstan-guide/parallelization.Rmd:91]: Thanks!

File: src/cmdstan-guide/parallelization.Rmd [hunk 3]
```diff
@@ -1,37 +1,173 @@
...
+### Dependencies
+
+OpenCL is supported on most modern CPUs and GPUs. In order to run OpenCL-enabled Stan models,
+an OpenCL runtime for the target device must be installed. This subsection lists installation
+instructions for OpenCL runtimes of the commonly-found devices.
+
+In order to check if any OpenCL-enabled device and its runtime is already present use the
+`clinfo` tool. On Linux, `clinfo` can typically be installed with the default package manager
+(for example `sudo apt-get install clinfo` on Ubuntu). For Windows, pre-built `clinfo` binary
+can be found [here](https://github.com/Oblomov/clinfo#windows-support).
```
Comment (bbbales2) [inline src/cmdstan-guide/parallelization.Rmd:111]: Is this really the clinfo Windows users use? Is there not a Microsoft one or something? From the Github: "The application can usually be built in Windows too"

File: src/cmdstan-guide/parallelization.Rmd [hunk 4]
```diff
@@ -1,37 +1,173 @@
...
+#### AMD GPU
+
+Linux:
+
+Install `Radeon Software for Linux` available [here](https://www.amd.com/en/support/kb/release-notes/rn-amdgpu-unified-linux-20-40).
+
+Windows: 
+
+We recommend installing the open source [OCL-SDK](https://github.com/GPUOpen-LibrariesAndSDKs/OCL-SDK/releases).
+#### AMD CPU
```
Comment (bbbales2) [inline src/cmdstan-guide/parallelization.Rmd:141]: Add a space before cause markdown can be picky.

File: src/cmdstan-guide/parallelization.Rmd [hunk 5]
```diff
@@ -1,37 +1,173 @@
...
+```
+or on Windows
+```
+mingw32-make STAN_OPENCL=true path/to/model.exe
+```
+The flag can also be written to the `make/local` file.
+
+### Running
+
+The Stan model compiled with `STAN_OPENCL` must also be supplied the OpenCL platform and device IDs.
```
Comment (bbbales2) [inline src/cmdstan-guide/parallelization.Rmd:165]: My computer does defaults and seems to get the right thing.

File: src/stan-users-guide/parallelization.Rmd [hunk 6]
```diff
@@ -687,3 +699,62 @@ function if they are rectangular.
 If the outputs are of varying sizes, then there will have to be some way
 to convert it back to a usable form based on the input, because there
 is no way to directly return sizes or a ragged structure.
+
+## OpenCL
+
+OpenCL (Open Computing Language) is a framework that enables writing programs that
+execute across heterogeneous platforms. An OpenCL program can be run on CPUs, GPUs,
+DSPs or FPGAs. In order to run OpenCL programs, an OpenCL runtime be installed on
```
Comment (bbbales2) [inline src/stan-users-guide/parallelization.Rmd:707]: Has anyone run Stan on a DSP or FPGA? If not let's just say OpenCL lets Stan run on GPUs.

File: src/stan-users-guide/parallelization.Rmd [hunk 7]
```diff
@@ -687,3 +699,62 @@ function if they are rectangular.
...
+they target system.
```
Comment (bbbales2) [inline src/stan-users-guide/parallelization.Rmd:708]: they -> the

File: src/stan-users-guide/parallelization.Rmd [hunk 8]
```diff
@@ -687,3 +699,62 @@ function if they are rectangular.
...
+- poisson_log_glm_lpmf*
+- rayleigh_lpdf
+- scaled_inv_chi_square_lpdf
+- skew_normal_lpdf
+- std_normal_lpdf
+- student_t_lpdf
+- uniform_lpdf
+- weibull_lpdf
+
+* GLM functions are used in cases where X is not a `row_vector`.
```
Comment (bbbales2) [inline src/stan-users-guide/parallelization.Rmd:757]: Is this still true?

File: src/stan-users-guide/parallelization.Rmd [hunk 9]
```diff
@@ -687,3 +699,62 @@ function if they are rectangular.
...
+
+### Example
```
Comment (bbbales2) [inline src/stan-users-guide/parallelization.Rmd:759]: Is there an example following?

File: src/stan-users-guide/parallelization.Rmd [hunk 10]
```diff
@@ -21,12 +29,16 @@ The actual speedup gained from using these functions will depend on
 many details. It is strongly recommended to only parallelize the
 computationally most expensive operations in a Stan
 program. Oftentimes this is the evaluation of the log likelihood for
-the observed data. Since only portions of a Stan program will run in
+the observed data. When in doubt which parts of the model is the most
```
Comment (bbbales2) [inline src/stan-users-guide/parallelization.Rmd:32]: > When in doubt which parts

"When it is not clear which parts"

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Comments!

File: src/cmdstan-guide/parallelization.Rmd [hunk 11]
```diff
@@ -1,37 +1,173 @@
...
-parallelization on an OpenCL (GPU) machine:
+## Multi-threading with TBB
 
+In order to exploit multi-threading in a Stan model, the models must be 
+rewritten to use the `reduce_sum` and `map_rect` functions. For instructions
+on how to rewrite Stan models to use these functions see [Stan's User guide chapter on parallelization](https://mc-stan.org/docs/stan-users-guide/parallelization-chapter.html), [the `reduce_sum` case study](https://mc-stan.org/users/documentation/case-studies/reduce_sum_tutorial.html) by Ben Bales and Sebastian Weber or the [Multithreading and Map-Reduce tutorial](https://github.com/rmcelreath/cmdstan_map_rect_tutorial) written by Richard McElreath.
+
+### Compiling
+Once a model is rewritten to use the above-mentioned functions, the model
+must be compiled with the `STAN_THREADS` makefile flag. The flag can be
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply src/cmdstan-guide/parallelization.Rmd:20]: Sure.

File: src/cmdstan-guide/parallelization.Rmd [hunk 12]
```diff
@@ -1,37 +1,173 @@
...
+implementations are [MPICH](https://www.mpich.org/) and [OpenMPI](https://www.open-mpi.org/).
+
+### Compiling
+
+Once a model is rewritten to use `map_rect`, additional makefile flags 
+must be written to the `make/local`. These are:
+
+- `STAN_MPI`: Enables the use of MPI with Stan if `TRUE`.
+- `CXX`: The name of the MPI C++ compiler wrapper. Typically `"mpicxx"`.
+- `TBB_CXX_TYPE`: The C++ compiler the MPI wrapper wraps. Typically `"gcc"` on Linux and `"clang"` on macOS.
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply src/cmdstan-guide/parallelization.Rmd:74]: These are CXX types and those are gcc and clang (tbb makefiles call gcc/clang for some files).

File: src/cmdstan-guide/parallelization.Rmd (hunk 3 above)
Comment (rok-cesnovar) [reply src/cmdstan-guide/parallelization.Rmd:111]: Yes, this is the Windows clinfo equivalent. 

There are all sorts of heavyweight GUI apps that also list the IDs, but this is a single exe file and straightforward.

Comment (rok-cesnovar) [reply src/cmdstan-guide/parallelization.Rmd:111]: You can download the built executable, which is what this links too (bottom of README).

File: src/stan-users-guide/parallelization.Rmd (hunk 8 above)
Comment (rok-cesnovar) [reply src/stan-users-guide/parallelization.Rmd:757]: Yes.

File: src/stan-users-guide/parallelization.Rmd (hunk 9 above)
Comment (rok-cesnovar) [reply src/stan-users-guide/parallelization.Rmd:759]: Yes, i wanted feedback if the rest makes sense before I embark on that.

File: src/cmdstan-guide/parallelization.Rmd (hunk 5 above)
Comment (rok-cesnovar) [reply src/cmdstan-guide/parallelization.Rmd:165]: Added a line on defaults working ok.

File: src/stan-users-guide/parallelization.Rmd (hunk 6 above)
Comment (rok-cesnovar) [reply src/stan-users-guide/parallelization.Rmd:707]: I have a labmate experimenting with FPGA and some basic Stan code, but yeah, I agree.

Comment (bbbales2) [reply src/stan-users-guide/parallelization.Rmd]: Oh interesting. That's more of a feature for us as OpenCL users tho -- as user facing docs, there isn't a supported way to run Stan on your FPGA as of now.

Comment (rok-cesnovar) [reply src/stan-users-guide/parallelization.Rmd]: Indeed.

File: src/stan-users-guide/parallelization.Rmd (hunk 8 above)
Comment (bbbales2) [reply src/stan-users-guide/parallelization.Rmd:756]: What does this bit mean though?

Comment (bbbales2) [reply src/stan-users-guide/parallelization.Rmd:756]: Like:

```
ordered_logistic_glm_lpmf*
```

GLM functions are used... But of course they're used, it's the glm function!

Comment (rok-cesnovar) [reply src/stan-users-guide/parallelization.Rmd:756]: With OpenCL is missing.

**Review by bbbales2 — APPROVED** [review]
I am happy with this. Pushed a couple changes. Merge it if you are happy.

**Pull request merged**